contracts/LogarithmicBuckets.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import "./DoubleLinkedList.sol";
+
+library LogarithmicBuckets {
+    using DoubleLinkedList for DoubleLinkedList.List;
+
+    struct BucketList {
+        mapping(uint256 => DoubleLinkedList.List) lists; // All the accounts.
+        mapping(address => uint256) indexOf;
+        uint256 maxIndex;
+    }
+
+    /// CONSTANTS ///
+
+    uint256 private constant LOGBASE = 2;
+
+    /// ERRORS ///
+
+    /// @notice Thrown when the address is zero at insertion.
+    error AddressIsZero();
+
+    /// INTERNAL ///
+
+    /// @notice Updates an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with value `_formerValue` or when `_id` is not in the `_heap` with `_formerValue` equal to 0.
+    function update(
+        BucketList storage _buckets,
+        address _id,
+        uint256 _formerValue,
+        uint256 _newValue
+    ) internal {
+        uint96 formerValue = SafeCast.toUint96(_formerValue);
+        uint96 newValue = SafeCast.toUint96(_newValue);
+
+        if (formerValue != newValue) {
+            if (formerValue != 0) remove(_buckets, _id);
+            if (newValue != 0) insert(_buckets, _id, newValue);
+        }
+    }
+
+    /// PRIVATE ///
+
+    function insert(
+        BucketList storage _buckets,
+        address _id,
+        uint96 _value
+    ) internal {
+        // `_buckets` cannot contain the 0 address.
+        if (_id == address(0)) revert AddressIsZero();
+        uint256 bucketIndex = computeBucketIndex(_value);
+        _buckets.lists[bucketIndex].insertTail(_id, _value);
+        _buckets.indexOf[_id] = bucketIndex;
+        if (bucketIndex > _buckets.maxIndex) _buckets.maxIndex = bucketIndex;
+    }
+
+    /// @notice Removes an account in the `_buckets`.
+    /// @param _buckets The buckets to modify.
+    /// @param _id The address of the account to remove.
+    function remove(BucketList storage _buckets, address _id) internal {
+        uint256 index = _buckets.indexOf[_id];
+        // Swap the last account and the account to remove, then pop it.
+        _buckets.lists[index].remove(_id);
+        delete _buckets.indexOf[_id];
+
+        while (_buckets.lists[index].head == address(0) && index > 0) {
+            index -= 1;
+            _buckets.maxIndex = index;
```

Comment:
Should be stored in a temporary variable to save gas and update the storage variable only at the end.